{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load all libraries\n",
    "import pandas as pd\n",
    "import ncbi.datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Start a Datasets gene API instance\n",
    "api_client = ncbi.datasets.ApiClient()\n",
    "ds_gene_instance = ncbi.datasets.GeneApi(api_client)\n",
    "\n",
    "# Retrieve gene summaries for the three genes using NCBI Gene IDs\n",
    "gene_summary = ds_gene_instance.gene_metadata_by_id([100699762])\n",
    "\n",
    "# Look up the symbols, chromosome number and SwissProt accession for each gene ,84347,140724,140725,140726,140727,140728,140729,501675,688285\n",
    "def report_on_gene_descriptors(gene_summary, leader='\\t', report_errors=True):\n",
    "    if report_errors:\n",
    "        for message in gene_summary.messages or []:\n",
    "            print(f'{leader}Error for: ({\",\".join(message.error.invalid_identifiers)})')\n",
    "            print(f'{leader}{leader}Reason: ({message.error.reason})')\n",
    "\n",
    "    if not gene_summary.genes:\n",
    "        print(f'{leader}No genes found')\n",
    "        return\n",
    "\n",
    "    for gene in map(lambda g: g.gene, gene_summary.genes):\n",
    "        print(f'{gene.symbol} (GeneID: {gene.gene_id}), Chromosome: {gene.chromosomes}, SwissProt: {gene.swiss_prot_accessions}')\n",
    "\n",
    "report_on_gene_descriptors(gene_summary)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def get_acc(species):\n",
    "    acc = []\n",
    "    df = pd.read_excel(\"C:\\\\Users\\\\junjie\\\\Desktop\\\\lab\\\\work\\\\Myfile\\\\All_species.xlsx\", sheet_name=species)\n",
    "    for i in df[\"accession_number\"]:\n",
    "        acc.append(i)\n",
    "    return acc\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#search location by gene id and protein accession number\n",
    "\n",
    "acc = [\"XP_003965054.2\",\"XP_003978708.1\"]\n",
    "geneID = [101064909,101061612]\n",
    "\n",
    "def searchlocation(acc, geneID):\n",
    "    api_client = ncbi.datasets.ApiClient()\n",
    "    ds_gene_instance = ncbi.datasets.GeneApi(api_client)\n",
    "    gene_summary = ds_gene_instance.gene_metadata_by_id([geneID])\n",
    "\n",
    "    for i in gene_summary.genes:\n",
    "        chromosome = (i.gene.chromosomes[0])\n",
    "        complement = i.gene.orientation\n",
    "        symbol = i.gene.symbol\n",
    "        cdna = \"\"\n",
    "        start = \"\"\n",
    "        stop = \"\"\n",
    "        complement = \"\"\n",
    "        accession_version = \"\"\n",
    "        cdnastart = \"\"\n",
    "        cdnastop = \"\"\n",
    "        exon = []\n",
    "        exonstart = 0\n",
    "        \n",
    "        if not i.gene.transcripts:\n",
    "            pass\n",
    "        else:\n",
    "            for l in i.gene.transcripts:\n",
    "                if not l.protein:\n",
    "                    pass\n",
    "                elif l.protein.accession_version == acc:\n",
    "                    accession_version = i.gene.transcripts[i.gene.transcripts.index(l)].genomic_range.accession_version\n",
    "                    cdna = i.gene.transcripts[i.gene.transcripts.index(l)].accession_version\n",
    "\n",
    "\n",
    "                    cdnastart = (i.gene.transcripts[i.gene.transcripts.index(l)].cds.range[0].begin)\n",
    "                    cdnastop = (i.gene.transcripts[i.gene.transcripts.index(l)].cds.range[0].end)\n",
    "\n",
    "\n",
    "                    start = (i.gene.transcripts[i.gene.transcripts.index(l)].genomic_range.range[0]['begin'])\n",
    "                    stop = (i.gene.transcripts[i.gene.transcripts.index(l)].genomic_range.range[0]['end'])\n",
    "                    complement = (i.gene.transcripts[i.gene.transcripts.index(l)].genomic_range.range[0]['orientation'])\n",
    "\n",
    "                    exon = [len(i.gene.transcripts[i.gene.transcripts.index(l)].genomic_locations[0].exons)]\n",
    "                    for num in range (len(i.gene.transcripts[i.gene.transcripts.index(l)].genomic_locations[0].exons)):\n",
    "                        exon.append(i.gene.transcripts[i.gene.transcripts.index(l)].genomic_locations[0].exons[num])\n",
    "                    exonstart = int(exon[1]['end'])-int(exon[1]['begin'])-int(cdnastart)\n",
    "                    tempexon = int(exonstart)\n",
    "                    print(tempexon)\n",
    "\n",
    "                else:\n",
    "                    pass\n",
    "\n",
    "        if str(complement)==\"minus\":\n",
    "            location = symbol + \",\" + acc + \",\\\"Chromosome: \" + chromosome + \"; \" + accession_version + \"(\" + start + \"..\" + stop + \", complement)\\\",\" + cdna + \", \" + cdnastart + \", \" + cdnastop + \", \" + str(exonstart)\n",
    "            for num in range (2,len(exon)-1):\n",
    "                tempexon += int(exon[num]['end'])-int(exon[num]['begin'])\n",
    "                location += \",\" + str(tempexon)\n",
    "        elif str(complement) != \"minus\":\n",
    "            location = symbol +  \",\" + acc + \",Chromosome: \" + chromosome + \"; \" + accession_version + \"(\" + start + \"..\" + stop + \"),\" + cdna + \",\" + cdnastart + \", \" + cdnastop + \", \" + str(exonstart)\n",
    "            for num in range (2,len(exon)-1):\n",
    "                tempexon += int(exon[num]['end'])-int(exon[num]['begin'])\n",
    "                location += \",\" + str(tempexon)\n",
    "        return location\n",
    "\n",
    "for i in range(len(acc)):\n",
    "    print(searchlocation(acc[i], geneID[i]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def writecsvtxt(acc,geneID):\n",
    "    temp = []\n",
    "    for i in range(len(acc)):\n",
    "        temp.append(searchlocation(acc[i], geneID[i]))\n",
    "    return temp\n",
    "\n",
    "\n",
    "txt = writecsvtxt(acc,geneID)\n",
    "\n",
    "for i in txt:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### output to csv\n",
    "def outputcsv(txt):\n",
    "    file = open(\"output.csv\",'w')\n",
    "    file.write(\"protein,accession,location,cDNAs,begin,end,exon_boundary1,exon_boundary2,exon_boundary3,exon_boundary4,exon_boundary5,exon_boundary6,exon_boundary7,exon_boundary8\\n\")\n",
    "    for i in range(len(txt)):\n",
    "        if i != len(txt)-1:\n",
    "            file.write(txt[i] + \"\\n\")\n",
    "        else:\n",
    "            file.write(txt[i])\n",
    "    file.close()\n",
    "outputcsv(txt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### get begin and end\n",
    "def get_start_end():\n",
    "    df = pd.read_csv(\"output.csv\")\n",
    "    return list(df['begin']), list(df['end']),list(df['cDNAs'])\n",
    "\n",
    "start = get_start_end()[0]\n",
    "stop = get_start_end()[1]\n",
    "cDNAs = get_start_end()[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "####get sequence\n",
    "\n",
    "import re\n",
    "import pandas as pd\n",
    "from urllib.request import urlopen\n",
    "import os\n",
    "\n",
    "\n",
    "def AccesiontoID(accession_num):\n",
    "    url = \"https://www.ncbi.nlm.nih.gov/protein/\"+str(accession_num)+\"?report=gilist&log$=seqview&format=text\"\n",
    "    df = urlopen(url)\n",
    "    for lines in df:\n",
    "        lines = lines.decode(\"utf-8\")\n",
    "        if \"<pre>\" in lines:\n",
    "            id = lines[5:].rstrip()\n",
    "    return id\n",
    "\n",
    "def get_sequence(geneid,start,stop):\n",
    "    temp = \"\"\n",
    "    url = \"https://www.ncbi.nlm.nih.gov/sviewer/viewer.cgi?tool=portal&save=file&log$=seqview&db=nuccore&report=fasta&id=\" + geneid + \"&from=\" + str(start) + \"&to=\" + str(stop) + \"&extrafeat=null&conwithfeat=on&hide-cdd=on\"\n",
    "    file = urlopen(url)\n",
    "    for line in file:\n",
    "        temp += line.decode(\"utf-8\")\n",
    "    return temp\n",
    "\n",
    "for i in range(len(cDNAs)):\n",
    "    txt = get_sequence(AccesiontoID(cDNAs[i]),start[i],stop[i])\n",
    "    print(txt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"C:\\\\Users\\\\junjie\\\\Desktop\\\\lab\\\\work\\\\Myfile\\\\New folder\\\\\"\n",
    "\n",
    "def createfilename(txt):\n",
    "    path = \"C:\\\\Users\\\\junjie\\\\Desktop\\\\lab\\\\work\\\\Myfile\\\\New folder\\\\\"\n",
    "    pattern = \"(?<=\\\\s).+\"\n",
    "    temp = re.findall(pattern,txt)\n",
    "    temp = re.sub(\"[;:]\",\"\",temp[0])\n",
    "    return path + temp + \".txt\"\n",
    "\n",
    "def writetxt(path,txt):\n",
    "    with open (path,\"w\") as obj:\n",
    "        obj.write(txt)\n",
    "\n",
    "\n",
    "temp = \"\"\n",
    "for i in range(len(cDNAs)):\n",
    "    txt = get_sequence(AccesiontoID(cDNAs[i]),start[i],stop[i])\n",
    "    temp+=txt\n",
    "    writetxt(createfilename(txt),txt)\n",
    "\n",
    "writetxt(\"C:\\\\Users\\\\junjie\\\\Desktop\\\\lab\\\\work\\\\Myfile\\\\New folder\\\\\" + \"all.txt\",temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pdf2image import convert_from_path\n",
    "pages = convert_from_path('C:\\\\Users\\\\junjie\\\\Desktop\\\\推甄\\\\陽明交通大學(陽明)\\\\自傳signed.pdf', 1000)\n",
    "\n",
    "for page in pages:\n",
    "    page.save('C:\\\\Users\\\\junjie\\\\Desktop\\\\推甄\\\\陽明交通大學(陽明)\\\\自傳signed.jpg', 'JPEG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#search location by gene id and protein accession number\n",
    "\n",
    "acc = [\"XP_003456412.1\",\"XP_005463339.1\",\"XP_005463337.1\"]\n",
    "geneID = [100698262,100696764,100696764]\n",
    "\n",
    "for i in range(len(acc)):\n",
    "    print(searchlocation(acc[i], geneID[i]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene structure and transmembrane domains analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read data from excel\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "dataf = pd.read_excel(\"data\\\\transmembrane\\\\Book1.xlsx\", sheet_name = \"Sheet1\")\n",
    "textname = []\n",
    "nuclen = []\n",
    "aalen = []\n",
    "exonboun = []\n",
    "trans = []\n",
    "templist = []\n",
    "\n",
    "for i in range(0,len(dataf[\"species\"])):\n",
    "    textname.append(dataf[\"species\"][i]+\" \"+dataf[\"protein\"][i])\n",
    "    templen = int(dataf[\"end\"][i])-int(dataf[\"begin\"][i])+1\n",
    "    nuclen.append(templen)\n",
    "    aalen.append(templen/3)\n",
    "    for j in range(1,6):\n",
    "        templist.append(dataf[\"exon_boundary\"+str(j)][i])\n",
    "    templist = [x for x in templist if x ==x ]\n",
    "    exonboun.append(templist)\n",
    "    templist = []\n",
    "    for k in range(1,6):\n",
    "        templist.append(dataf[\"trans\"+str(k)+\"_start\"][i])\n",
    "        templist.append(dataf[\"trans\"+str(k)+\"_end\"][i])\n",
    "    templist = [x for x in templist if x ==x ]\n",
    "    trans.append(templist)\n",
    "    templist = []\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create figure\n",
    "\n",
    "from PIL import Image, ImageDraw, ImageFont\n",
    "\n",
    "newImage = Image.new('RGBA', (2000,2500), 'White')\n",
    "drawObj = ImageDraw.Draw(newImage)\n",
    "\n",
    "drawstart = 450\n",
    "recx1 = 5\n",
    "recx2 = 15\n",
    "recy1 = 20\n",
    "recy2 = 40\n",
    "texty = 15\n",
    "triy1 = 50\n",
    "triy2 = triy1+30\n",
    "gap = 180\n",
    "adjust = 1.5\n",
    "\n",
    "fontInfo = ImageFont.truetype('C:\\Windows\\Fonts\\calibri.ttf',45)\n",
    "\n",
    "for item in range(0,len(dataf)):\n",
    "\n",
    "    drawObj.text((20,texty), textname[item], fill='Black', font=fontInfo)\n",
    "\n",
    "    drawObj.rectangle((drawstart,recy1,drawstart+adjust*aalen[item],recy2),fill='lightgrey')\n",
    "\n",
    "    for i in range (0,len(trans[item]),2):\n",
    "        drawObj.rectangle((drawstart+round(adjust*trans[item][i]),recx1,drawstart+round(adjust*trans[item][i+1]),recx2),fill='Black')\n",
    "    \n",
    "    for i in range (0,len(exonboun[item])):\n",
    "        drawObj.polygon([(450+(round(exonboun[item][i]/nuclen[item]*adjust*aalen[item]))-10,triy2),(450+(round(exonboun[item][i]/nuclen[item]*adjust*aalen[item])),triy1),(450+(round(exonboun[item][i]/nuclen[item]*adjust*aalen[item]))+10,triy2)],fill='Black')\n",
    "\n",
    "    texty += gap\n",
    "    recy1 += gap\n",
    "    recy2 += gap\n",
    "    recx1 += gap\n",
    "    recx2 += gap\n",
    "    triy1 += gap\n",
    "    triy2 += gap\n",
    "\n",
    "\n",
    "drawObj.polygon([(950-10,texty+35),(950,texty+5),(950+10,texty+35)],fill='Black')\n",
    "drawObj.text((1000,texty), \": Exon boundaries\", fill='Black', font=fontInfo, align='right')\n",
    "\n",
    "drawObj.rectangle((925,texty+90,970,texty+100),fill='Black')\n",
    "drawObj.text((1000,texty+70), \": Predicted transmembrane segments\", fill='Black', font=fontInfo, align='right')\n",
    "\n",
    "newImage.save(\"structureout.png\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Syntenic Block analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read data from csv\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "class Chromosome:\n",
    "    def __init__(self, species, chr):\n",
    "        self.species = species\n",
    "        self.chr = chr\n",
    "\n",
    "    def read_csv(self):\n",
    "        dataf = pd.read_csv(\"data\\\\syntenic\\\\\"+self.species + \" C\" + str(self.chr) + \".csv\", skiprows=[0])\n",
    "        return dataf\n",
    "\n",
    "def get_df(cdict = {\"Homo sapiens\":[16,17,19,22], \"Mus musculus\":[7,11,15,16]}):\n",
    "    specieslist = []\n",
    "    temp = []\n",
    "    count = 1\n",
    "    for i in cdict:\n",
    "        for j in cdict[i]:\n",
    "            exec(f\"c{count} = Chromosome(i, j)\")\n",
    "            exec(f\"df{count} = c{count}.read_csv()\")\n",
    "            exec(f\"temp.append(df{count})\")\n",
    "            count += 1\n",
    "        specieslist.append(temp)\n",
    "        temp = []\n",
    "    return specieslist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get neighbour gene\n",
    "\n",
    "def get_neighbour(dataf,n):\n",
    "    address = []\n",
    "    distance = n\n",
    "    neighbour_gene = []\n",
    "    for i in range(len(dataf[\"Gene symbol\"])):\n",
    "        if \"CACNG\" in dataf[\"Gene symbol\"][i].upper():\n",
    "            for j in range (0,distance+1):\n",
    "                if i-j>=0:address.append(i-j)\n",
    "                if i+j<len(dataf[\"Gene symbol\"]):address.append(i+j)\n",
    "                \n",
    "        elif \"TMEM114\" in dataf[\"Gene symbol\"][i].upper():\n",
    "            for j in range (0,distance+1):\n",
    "                if i-j>=0:address.append(i-j)\n",
    "                if i+j<len(dataf[\"Gene symbol\"]):address.append(i+j)\n",
    "        elif \"TMEM235\" in dataf[\"Gene symbol\"][i].upper():\n",
    "            for j in range (0,distance+1):\n",
    "                if i-j>=0:address.append(i-j)\n",
    "                if i+j<len(dataf[\"Gene symbol\"]):address.append(i+j)\n",
    "\n",
    "    address=sorted(set(address))\n",
    "\n",
    "    for i in address:\n",
    "        neighbour_gene.append(dataf[\"Gene symbol\"][i])\n",
    "    return neighbour_gene\n",
    "\n",
    "def get_neighbourlist(specieslist,n):\n",
    "    temp = []\n",
    "    neighbour_gene = []\n",
    "    for i in range(len(specieslist)):\n",
    "        for j in range(len(specieslist[i])):\n",
    "            temp.append(get_neighbour(specieslist[i][j],n))\n",
    "        neighbour_gene.append(temp)\n",
    "        temp = []   \n",
    "    return neighbour_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "cdict = {\"Homo sapiens\":[16,17,22,19], \"Mus musculus\":[16,11,15,7], \"Canis lupus familiaris\":[6,9,10,1]}\n",
    "common_name = [\"Human\", \"Mouse\", \"Dog\"]\n",
    "neighbour_gene = get_neighbourlist(get_df(cdict),10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#count score\n",
    "\n",
    "from itertools import chain\n",
    "from collections import Counter\n",
    "\n",
    "neighbour_gene_list = list(chain.from_iterable(list(chain.from_iterable(neighbour_gene))))\n",
    "for i in range(len(neighbour_gene_list)):\n",
    "    neighbour_gene_list[i] = neighbour_gene_list[i].upper()\n",
    "\n",
    "scoredict = Counter(neighbour_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#arrange output data\n",
    "\n",
    "def filter_score(scoredict,score):\n",
    "    templist = []\n",
    "    for i in scoredict:\n",
    "        if scoredict[i] >= score:\n",
    "            templist.append(i)\n",
    "    return templist\n",
    "\n",
    "def output():\n",
    "    temp = []\n",
    "    templist = []\n",
    "    templist1 = []\n",
    "    for i in range (len(get_df(cdict))):\n",
    "        for j in range (len(get_df(cdict)[i])):\n",
    "            for k in get_df(cdict)[i][j][\"Gene symbol\"]:\n",
    "                if k.upper() in filter_score(scoredict,3):\n",
    "                    temp.append(k)\n",
    "            templist.append(temp)\n",
    "            temp = []\n",
    "        templist1.append(templist)\n",
    "        templist = []\n",
    "    return templist1\n",
    "\n",
    "outputdata = output()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CACNG1': [400, 1720, 1400, 1270, 2400, 2170], 'CACNG2': [400, 2970, 1400, 2570, 2400, 3020], 'CACNG3': [400, 920, 1400, 3870, 2400, 570], 'CACNG4': [400, 1670, 1400, 1320, 2400, 2220], 'CACNG5': [400, 1620, 1400, 1370, 2400, 2270], 'CACNG6': [400, 3820, 1400, 3420, 2400, 3820], 'CACNG7': [400, 3720, 1400, 3320, 2400, 3920], 'CACNG8': [400, 3770, 1400, 3370, 2400, 3870], 'TMEM114': [400, 420, 1400, 420, 2400, 1070], 'TMEM235': [400, 2220, 1400, 1820, 2400, 1670]}\n"
     ]
    }
   ],
   "source": [
    "# draw\n",
    "\n",
    "from PIL import Image, ImageDraw, ImageFont\n",
    "\n",
    "chro = []\n",
    "\n",
    "for i in cdict:\n",
    "    for j in cdict[i]:\n",
    "     chro.append(j)\n",
    "\n",
    "startpointx = 250\n",
    "startpointy = 400\n",
    "textx1 = 200\n",
    "texty1 = 50\n",
    "textx2 = 0\n",
    "texty2 = 400\n",
    "\n",
    "fontInfo1 = ImageFont.truetype('C:\\Windows\\Fonts\\calibri.ttf',150)\n",
    "fontInfo2 = ImageFont.truetype('C:\\Windows\\Fonts\\calibri.ttf',50)\n",
    "\n",
    "newImage = Image.new('RGBA', (10000,10500), 'White')\n",
    "drawObj = ImageDraw.Draw(newImage)\n",
    "# drawObj.rectangle((150,startpointx+50,100,startpointx+100),fill='black')\n",
    "# drawObj.rectangle((150,startpointx+100,100,startpointx+150),outline='black')\n",
    "# startpointx+=100\n",
    "tempindex = 0\n",
    "\n",
    "for i in range (len(outputdata)):\n",
    "    drawObj.text((textx1, texty1),common_name[i],fill=\"black\",font=fontInfo1)\n",
    "    textx1 += 1000\n",
    "    for j in range (len(outputdata[i])):\n",
    "        # drawObj.text((startpointx-50,texty), \"asdf\", fill='Black', font=fontInfo)\n",
    "        drawObj.text((startpointx, texty2-100),\"Chromosome \"+str(chro[tempindex]),fill=\"black\",font=fontInfo2)\n",
    "        drawObj.rectangle((startpointx,startpointy,startpointx+30,len(outputdata[i][j])*50+startpointy),fill='black')\n",
    "        startpointy+=400+len(outputdata[i][j])*50\n",
    "        texty2 += 400+len(outputdata[i][j])*50\n",
    "        tempindex += 1\n",
    "    startpointy = 400\n",
    "    texty2 = 400\n",
    "    startpointx += 1000\n",
    "\n",
    "drawline={\"CACNG1\":[],\"CACNG2\":[],\"CACNG3\":[],\"CACNG4\":[],\"CACNG5\":[],\"CACNG6\":[],\"CACNG7\":[],\"CACNG8\":[], \"TMEM114\":[], \"TMEM235\":[]}\n",
    "startpointx = 400\n",
    "texty2,startpointy = 400,400\n",
    "for i in range (len(outputdata)):\n",
    "    for j in range (len(outputdata[i])):\n",
    "        for k in range (len(outputdata[i][j])):\n",
    "            # drawObj.text((textx2, texty2),outputdata[i][j][k],fill=\"black\",font=fontInfo2)\n",
    "            if \"CACNG\" in outputdata[i][j][k].upper():\n",
    "                drawObj.text((startpointx-100, texty2),outputdata[i][j][k].upper(),fill=\"red\",font=fontInfo2)\n",
    "                drawline[outputdata[i][j][k].upper()].append(startpointx)\n",
    "                drawline[outputdata[i][j][k].upper()].append(texty2+20)\n",
    "            elif \"TMEM114\" in outputdata[i][j][k].upper():\n",
    "                drawObj.text((startpointx-100, texty2),outputdata[i][j][k].upper(),fill=\"red\",font=fontInfo2)\n",
    "                drawline[outputdata[i][j][k].upper()].append(startpointx)\n",
    "                drawline[outputdata[i][j][k].upper()].append(texty2+20)\n",
    "            elif \"TMEM235\" in outputdata[i][j][k].upper():\n",
    "                drawObj.text((startpointx-100, texty2),outputdata[i][j][k].upper(),fill=\"red\",font=fontInfo2)\n",
    "                drawline[outputdata[i][j][k].upper()].append(startpointx)\n",
    "                drawline[outputdata[i][j][k].upper()].append(texty2+20)\n",
    "            else:\n",
    "                drawObj.text((startpointx-100, texty2),outputdata[i][j][k].upper(),fill=\"black\",font=fontInfo2)\n",
    "\n",
    "            texty2+=50\n",
    "        startpointy+=400+len(outputdata[i][j])*50\n",
    "        texty2 = startpointy\n",
    "    texty2,startpointy = 400,400\n",
    "    startpointx += 1000\n",
    "\n",
    "\n",
    "print(drawline)\n",
    "for item in drawline:\n",
    "    for i in range (0,int(len(drawline[item])/2),2):\n",
    "        drawObj.line(((drawline[item][i]+120,drawline[item][i+1]),(drawline[item][i+2]-140,drawline[item][i+3])), fill=\"red\", width=3)\n",
    "    \n",
    "        \n",
    "newImage.save(\"out.png\")\n",
    "# newImage.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CACNG1': [400, 1700, 1400, 1250, 2400, 2150], 'CACNG2': [400, 2950, 1400, 2550, 2400, 3000], 'CACNG3': [400, 900, 1400, 3850, 2400, 550], 'CACNG4': [400, 1650, 1400, 1300, 2400, 2200], 'CACNG5': [400, 1600, 1400, 1350, 2400, 2250], 'CACNG6': [400, 3800, 1400, 3400, 2400, 3800], 'CACNG7': [400, 3700, 1400, 3300, 2400, 3900], 'CACNG8': [400, 3750, 1400, 3350, 2400, 3850], 'TMEM114': [400, 400, 1400, 400, 2400, 1050], 'TMEM235': [400, 2200, 1400, 1800, 2400, 1650]}\n",
      "400 1700\n",
      "1400 1250\n",
      "2400 2150\n",
      "400 2950\n",
      "1400 2550\n",
      "2400 3000\n",
      "400 900\n",
      "1400 3850\n",
      "2400 550\n",
      "400 1650\n",
      "1400 1300\n",
      "2400 2200\n",
      "400 1600\n",
      "1400 1350\n",
      "2400 2250\n",
      "400 3800\n",
      "1400 3400\n",
      "2400 3800\n",
      "400 3700\n",
      "1400 3300\n",
      "2400 3900\n",
      "400 3750\n",
      "1400 3350\n",
      "2400 3850\n",
      "400 400\n",
      "1400 400\n",
      "2400 1050\n",
      "400 2200\n",
      "1400 1800\n",
      "2400 1650\n"
     ]
    }
   ],
   "source": [
    "print(drawline)\n",
    "for item in drawline:\n",
    "    for i in range (0,len(drawline[item]),2):\n",
    "        print(drawline[item][i], drawline[item][i+1])"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "d2c1845aa41483b05395b2444e9cc1d1db12b4e36041ff17c25c8710311460d9"
  },
  "kernelspec": {
   "display_name": "Python 3.10.0 64-bit",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
